### ISSUE TITLE
Remove the containers and heaps dependency

### ISSUE THREAD

**Issue by harendra-kumar** [composewell/streamly#83]
We can possibly remove these dependencies as we use very little functionality from these which can be duplicated in the package itself.

**Comment by harendra-kumar** [comment]
Is it worth removing these? 

`containers` brings in `array` and `deepseq` as additional dependencies which are pretty basic themselves.

Even though `heaps` has no runtime dependencies it seems to be bringing Cabal and cabal-doctest as build time dependencies. Can those be removed from `heaps`?

**Comment by hasufell** [comment]
I noticed that it's impossible to find a build plan with a package, which:

* uses ghc-8.0
* uses streamly
* has doctests, which are scheduled to be built

The error message will look something like this

```
$ cabal new-build --enable-tests 
Resolving dependencies...
cabal: Could not resolve dependencies:
[__0] trying: hpath-0.9.2 (user goal)
[__1] rejecting: hpath:!test (constraint from config file, command line flag,
or user target requires opposite flag selection)
[__1] trying: hpath:*test
[__2] trying: doctest-0.16.2 (dependency of hpath *test)
[__3] trying: ghc-8.0.2/installed-8.0... (dependency of doctest)
[__4] next goal: streamly (dependency of hpath)
[__4] rejecting: streamly-0.7.0 (conflict: ghc =>
containers==0.5.7.1/installed-0.5..., streamly => containers>=0.5.8.2 && <0.7)
[__4] rejecting: streamly-0.6.1, streamly-0.6.0, streamly-0.5.2,
streamly-0.5.1, streamly-0.5.0, streamly-0.4.1, streamly-0.4.0,
streamly-0.3.0, streamly-0.2.1, streamly-0.2.0, streamly-0.1.2,
streamly-0.1.1, streamly-0.1.0 (conflict: hpath => streamly>=0.7)
[__4] fail (backjumping, conflict set: ghc, hpath, streamly)
After searching the rest of the dependency tree exhaustively, these were the
goals I've had most trouble fulfilling: ghc, doctest, hpath, streamly, base,
hpath:test
Try running with --minimize-conflict-set to improve the error message.
```

**Comment by harendra-kumar** [comment]
We will try to lower the version bounds of `containers` in the upcoming minor release this week.

**Comment by pranaysashank** [comment]
Lower bound on containers was bumped in 6ccac7d #274 . 

We are using `alterF` from `Data.Map.Strict` which is available only since 0.5.8.1, maybe we can instead provide it in the library.

**Comment by harendra-kumar** [comment]
heaps dependency from streamly-core will be removed when the deprecated modules are removed.
